src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -0,0 +1,574 @@
+      else a
+    case (Kon(tp1, ss1), Kon(tp2, ss2)) =>
+      if (!isEqualType(tp1, tp2)) a
+      else if (ss1.zip(ss2).exists(p => simplify(intersect(p._1, p._2)) == Empty)) a
+      else if (ss1.zip(ss2).forall((subspace _).tupled)) Empty
+      else
+        Or(
+          ss1.zip(ss2).map((minus _).tupled).zip(0 to ss2.length - 1).map {
+            case (ri, i) => Kon(tp1, ss1.updated(i, ri))
+          })
+    case (Const(v1, _), Const(v2, _)) =>
+      if (v1 == v2) Empty else a
+    case (Const(_, tp1), Typ(tp2, _)) =>
+      if (isSubType(tp1, tp2)) Empty else a
+    case (Const(_, _), _) => a
+    case (Typ(tp1, _), Const(_, tp2)) =>  // Boolean & Java enum
+      if (isSubType(tp2, tp1) && canDecompose(tp1))
+        minus(Or(partitions(tp1)), b)
+      else a
+    case (_, Const(_, _)) => a
+    case (Var(x, _), Var(y, _)) =>
+      if (x == y) Empty else a
+    case (Var(_, tp1), Typ(tp2, _)) =>
+      if (isSubType(tp1, tp2)) Empty else a
+    case (Var(_, _), _) => a
+    case (_, Var(_, _)) => a
+  }
+}
+
+/** Scala implementation of space logic */
+class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
+  import tpd._
+
+  /** Return the space that represents the pattern `pat`
+   *
+   *  If roundUp is true, approximate extractors to its type,
+   *  otherwise approximate extractors to Empty
+   */
+  def project(pat: Tree, roundUp: Boolean = true)(implicit ctx: Context): Space = pat match {
+    case Literal(c) => Const(c, c.tpe)
+    case _: BackquotedIdent => Var(pat.symbol, pat.tpe)
+    case Ident(_) | Select(_, _) =>
+      pat.tpe.stripAnnots match {
+        case tp: TermRef =>
+          if (pat.symbol.is(Enum))
+            Const(Constant(pat.symbol), tp)
+          else if (tp.underlyingIterator.exists(_.classSymbol.is(Module)))
+            Typ(tp.widenTermRefExpr.stripAnnots, false)
+          else
+            Var(pat.symbol, tp)
+        case tp => Typ(erase(tp), false)
+      }
+    case Alternative(trees) => Or(trees.map(project(_, roundUp)))
+    case Bind(_, pat) => project(pat)
+    case UnApply(_, _, pats) =>
+      if (pat.tpe.classSymbol.is(CaseClass))
+        Kon(pat.tpe.stripAnnots, pats.map(pat => project(pat, roundUp)))
+      else if (roundUp) Typ(pat.tpe.stripAnnots, false)
+      else Empty
+    case Typed(pat @ UnApply(_, _, _), _) => project(pat)
+    case Typed(expr, _) => Typ(erase(expr.tpe.stripAnnots), true)
+    case _ =>
+      Empty
+  }
+
+  /* Erase a type binding according to erasure rule */
+  def erase(tp: Type): Type = {
+    def doErase(tp: Type): Type = tp match {
+      case tp: RefinedType => erase(tp.parent)
+      case _ => tp
+    }
+
+    val origin = doErase(tp)
+    if (origin =:= defn.ArrayType) tp else origin
+  }
+
+  /** Is `tp1` a subtype of `tp2`?  */
+  def isSubType(tp1: Type, tp2: Type): Boolean = {
+    // expose is important due to type bounds
+    // check SI-9657 and tests/patmat/gadt.scala
+    tp1 <:< expose(tp2)
```

Comment:
It seems here we can't just reuse the existing type erasure by splitting or-types to or-spaces, e.g.  `M[Int | Boolean]` may not necessarily the same as `M[Int] | M[Boolean]`.